frontend/packages/console-shared/src/components/utils/with-resources.tsx

```diff
@@ -0,0 +1,129 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
+import { connect } from 'react-redux';
+import { Map as ImmutableMap } from 'immutable';
+
+import { Firehose, inject } from '@console/internal/components/utils';
+import { K8sKind, K8sResourceKindReference } from '@console/internal/module/k8s';
+
+type ResourceToPropsResult = {
+  value: {
+    [key: string]: any;
+  };
+};
+
+type ResourcesProps = {
+  loaderComponent?: React.ComponentType;
+  onError?(): void;
+  resourceMap: any;
+  resources?: any;
+  resourceToProps?(): ResourceToPropsResult;
+  children: React.ReactNode;
+};
+
+type ResourcesState = {
+  childrenProps: any;
+  errors: any[];
+  loaded: boolean;
+};
+
+function checkErrors(resources, onError): void {
+  if (resources.length > 0) {
+    if (onError) {
+      onError(resources);
+    }
+    resources.forEach(resource => {
+      const errorMessage = _.get(
+        resource.error,
+        'json.message',
+        `Error occured while loading ${resource.resourceConfig.resource.kind}`
+      );
+      const errorCode = _.get(resource.error, 'json.code', '');
+      const error = errorCode ? `${errorCode}: ${errorMessage}` : errorMessage;
+      console.warn(error); // eslint-disable-line
+    });
+  }
+}
+
+function getResourcesState({ resourceMap, resources = {}, resourceToProps = null }): ResourcesState {
+  const childrenProps = {};
+  const errors = [];
+  let loaded = true;
+
+  Object.keys(resourceMap).forEach(resourceKey => {
+    const resourceConfig = resourceMap[resourceKey];
+    const configResource = resourceConfig.resource;
+    const resource = _.get(resources, resourceKey);
+
+    if (resource) {
+      if (resource.loaded) {
+        childrenProps[resourceKey] = resource.data;
+      } else if (resourceConfig.required) {
+        loaded = false;
+      }
+
+      if (!resourceConfig.ignoreErrors && resource.loadError) {
+        childrenProps[resourceKey] = configResource.isList ? [] : {};
+        errors.push({ error: resource.loadError, resourceConfig });
+      }
+    } else {
+      // unknown resources (CRD not created in opeshift, etc..)
+      childrenProps[resourceKey] = configResource.isList ? [] : {};
+    }
+  });
+
+  return {
+    childrenProps: {
+      ...childrenProps,
+      ...(resourceToProps && loaded ? resourceToProps(childrenProps) : {}),
+    },
+    errors,
+    loaded,
+  };
+}
+
+const Resources = (props: ResourcesProps) => {
+  const [{ errors, loaded, childrenProps }, setResourcesState] = React.useState(getResourcesState(props));
+  const { onError, loaderComponent, children } = props;
+
+  React.useEffect(() => checkErrors(errors, onError), [errors]);
+  React.useEffect(() => setResourcesState(getResourcesState(props)), [props]);
```

Comment:
What is the second argument `[props]` for, please?

Reply:
IIUC, `getResourcesState(props)` will result in the same value as for the `useState()` default above. Can you please elaborate this flow?

Reply:
Second argument is the array of values that the effect depends on, the effect gets run only when one of the values changes. When an empty array is passed, the effect is only run on mount (same as componentDidMount). (https://reactjs.org/docs/hooks-reference.html#useeffect).

Original implementation used `static getDerivedStateFromProps` function to populate the state from props. First the state was initialized in constructor:
`this.state = Resources.getDerivedStateFromProps(props);`. To mimic this I am defining the `useState` hook and initialize the state by calling the function `getResourcesState(props)`.
Then to update the state on props change I am using useEffect hook which gets run only when props change.